Update how-to-build.md with instructions on installing cmake

ironage opened the pull request:
Fixes #2727

bdash commented on how-to-build.md:
```diff
@@ -2,10 +2,10 @@ This file explains how to build and install the Realm core library.
-tools. This includes a C/C++ compiler and GNU make. Realm is
-thoroughly tested with both GCC and Clang. It is known to work with
-GCC 4.2 and newer, as well as with Clang 3.0 and newer.
+To build the Realm core library, you need cmake and a standard set of
```
`s/cmake/CMake/`, here and in a number of other places.

bdash commented on how-to-build.md:
```diff
@@ -2,10 +2,10 @@ This file explains how to build and install the Realm core library.
+To build the Realm core library, you need cmake and a standard set of
+build tools. This includes a C/C++ compiler and a build system like
+GNU make. Realm is thoroughly tested with both GCC and Clang. It is
+known to work with GCC 4.2 and newer, as well as with Clang 3.0 and newer.
```
These compiler versions are quite outdated. I think GCC 4.9 is the oldest version we support. I'm not sure what the oldest version of Clang we test with is… maybe the one from Xcode 8.0?

kspangsege replied:
Yes, I believe we require GCC >= 4.9, because it is the earliest version that has sufficient C++14 support.

bdash commented on how-to-build.md:
```diff
@@ -18,30 +18,75 @@ each of our major platforms:
     sudo apt-get install build-essential
     sudo apt-get install libproc-dev
     sudo apt-get install libssl-dev
+    # To install cmake see Building CMake from Source below.
```
It might be worth making `Building CMake from Source` into a link?

bdash commented on how-to-build.md:
```diff
@@ -18,30 +18,75 @@ each of our major platforms:
+- Visual Studio 2017: https://www.visualstudio.com/
+- CMake version 3.6 or higher: https://cmake.org/download/
+
+### Building CMake from Source
```
Any reason to suggest building from source rather than downloading prebuilt binaries for Linux as well? That's what we do for CI, and it seems to work well for those purposes.

From object store's Dockerfile:
```sh
# Ensure a new enough version of CMake is available.
RUN cd /opt \
    && wget https://cmake.org/files/v3.7/cmake-3.7.2-Linux-x86_64.tar.gz \
        && tar zxvf cmake-3.7.2-Linux-x86_64.tar.gz

ENV PATH "/opt/cmake-3.7.2-Linux-x86_64/bin:$PATH"
```

jedelbo replied:
Or you could download the .sh version (https://cmake.org/files/v3.8/cmake-3.8.2-Linux-x86_64.sh) and then run
```
sudo sh cmake-3.8.2-Linux-x86_64.sh --prefix=/usr/local
```

ironage replied:
I like the idea of the shell script, that seems much simple and dependable. Thanks!

bdash commented on how-to-build.md:
```diff
@@ -18,30 +18,75 @@ each of our major platforms:
+it may not be available by default from your package manager. A reliable
+way to install a more recent version of cmake than the version your
+package manager provides is to build it from source as shown below.
+The following instructions are tested on fedora 24 and ubuntu 16.04.
```
s/fedora/Fedora/
s/ubuntu/Ubuntu/

ironage commented:
I will wait until #2728 is resolved before addressing feedback here since the versions and instructions will likely have to be updated again.

kspangsege commented on how-to-build.md:
```diff
@@ -18,30 +18,75 @@ each of our major platforms:
     sudo apt-get install build-essential
     sudo apt-get install libprocps0-dev
     sudo apt-get install libssl-dev
+    # To install cmake see Building CMake from Source below.
```
It is unlikely that these have a sufficiently new GCC by default. I say we drop these from the list.

ironage commented:
I've updated the docs for newer systems that we support including how to install cmake on those systems. I've also relaxed the minimum required CMake version to 3.4. I removed the section on building CMake from source since version 3.4 is available from all documented supported systems.

@bdash @jedelbo @kspangsege Please re-review.

bdash commented on how-to-build.md:
```diff
@@ -2,46 +2,58 @@ This file explains how to build and install the Realm core library.
+standard set of build tools. This includes a C/C++ compiler and a
+build system like GNU make. Realm is thoroughly tested with both GCC
+and Clang. It is known to work with GCC 4.9 and newer, as well as
+with Clang 8.0 and newer. Your compiler must support C++14.
```
"8.0" is part of a versioning scheme used by the version of Clang that Apple ships. The latest version using LLVM's versioning schema is 4.0. I don't know why Apple invented their own version numbering like this. I'm not sure what the best way to concisely clarify this is either.

ironage replied:
"... as well as with Clang 3.9 (Apple clang 8.0) and newer ..."
@bdash would that provide enough clarity without detracting from flow?

bdash replied:
Sounds good!

ironage replied:
Updated.
